### ISSUE TITLE
Make the Compiler and Language Service API Asynchronous

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#1857]
It makes very little sense that IO-bound operations (e.g. readFile, writeFile) are synchronous when we run on a platform whose touted strength is in writing IO-bound applications through asynchronous non-blocking APIs.

Compile times could probably be shaved down by a bit, especially for large projects with many files. Additionally, tooling experience should be as responsive as possible and avoid blocking in the event that IO takes place, enabling several optimizations to free memory on the managed side.

With #1664, this should be significantly easier.

This means that at the very least, the following would change to have asynchronous counterparts.

* `createSourceFile` (because of `/// <reference ... />` comments)
* `createProgram`
* ***all of the language service API***

**Comment by NoelAbrahams** [comment]
> Compile times could probably be shaved down by a bit, especially for large projects with many files. Additionally, tooling experience should be as responsive as possible and avoid blocking in the event that IO takes place

:+1: 

Does that mean you will ship and support a promise library?

**Comment by fdecampredon** [comment]
Honestly after the hell I have gone through to write brackets-typescript, and make languageService works in a webworker, I can only :+1:  :+1:  :+1:  this one.

**Comment by DanielRosenwasser** [comment]
> Does that mean you will ship and support a promise library? 

I think it'd still be bring-your-own-promise; whatever we use would probably not be an official endorsement of what to use.

**Comment by basarat** [comment]
Just FYI : people will still need to run in a webworker (or separate node process). Making `io` async alone is not the performance bottleneck : https://github.com/TypeStrong/atom-typescript/issues/36#issuecomment-72344996 Its just the sheer amount of *computations* that need to be performed.

**Comment by DanielRosenwasser** [comment]
Right - for an optimal experience, you really need to get some multithreading into the mix. In our case, we do some caching to avoid blocking IO operations, so this would help us in terms of reducing our memory footprint.

**Comment by vvakame** [comment]
:+1:

**Comment by basarat** [comment]
Definitely try https://atom.io/packages/atom-typescript again if you haven't before. Now I'm running a `node` (or `atom` on mac) instance just listening on stdin using this : https://github.com/TypeStrong/atom-typescript/blob/df110e7218815dc05be90102f46aad7bd2f04d29/lib/worker/messages.ts#L1-L50 

With a parent : https://github.com/TypeStrong/atom-typescript/blob/df110e7218815dc05be90102f46aad7bd2f04d29/lib/worker/messages.ts#L1-L50 

And a child: https://github.com/TypeStrong/atom-typescript/blob/df110e7218815dc05be90102f46aad7bd2f04d29/lib/worker/workerProcess.ts#L18-L36 

There are still cleanups to do but the experience is awesome. 

/cc (and thanks) @csnover

**Comment by basarat** [comment]
I've wrapped it up in an architecture that is *extremely simple* to understand / debug / extend. Also documented it (thanks to excellent support by @csnover) https://github.com/TypeStrong/atom-typescript/blob/master/CONTRIBUTING.md#architecture

**Comment by RyanCavanaugh** [comment]
This doesn't actually buy us much because we're not I/O-bound, and we don't have any I/O that can be easily parallelized (we produce a list of files basically serially). The impact on the compiler base would be very large for not any real perf gain.

**Comment by basarat** [comment]
> The impact on the compiler base would be very large for not any real perf gain.

Agreed. Anyone using the language service will *still* need to run it in the background even in this case as even after we have read all the FS files doing something like `getCompletionsAtPosition` takes about 81ms on my machine (CPU bound)

**Comment by CyrusNajmabadi** [comment]
The LS is very IO bound. Furthermore, the LS needs to be sync to be a good VS/Roslyn citizen. We must make the LS async for scalability :-)

    -- Cyrus
________________________________

**Comment by mhegazy** [comment]
I don't think we must :) we still can use the ts server, which can Handel most of the io. We should close this issue and file another one when we do perf and scalability work. Possibly with P2P work.

**Comment by CyrusNajmabadi** [comment]
@mhegazy This is already an issue for the VS case today.  Right now we end up having to load up all files on the VS side before we can even call into the script side to answer a question.  We do this so we have the data ready when the LS asks the host "what has changed between these files?"  And we have to do this, even though the LS might not ever ask that question (because the version hasn't changed).  If we were async through-and-through, we wouldn't have this problem.

We'd simply call into the script LS.  It would synchronize info with the host, and then the LS would just ask for the changes to files with different versions in an async manner.  This would be *much* better, and would provide a lot less stress on the managed side (for example, Roslyn could flush these files to their cache, instead of having to keep them all in memory.).

**Comment by DanielRosenwasser** [comment]
To add to what Cyrus said, I think it just generally be helpful to see some data regarding this before we came to any sort of conclusion - both on the compiler side as well as the LS side.

**Comment by mhegazy** [comment]
@CyrusNajmabadi I understand how the system works :). what I am saying is this should be done as part of other work items. We need to do scalability and performance work items, but these are not scheduled.. So all I am saying is close this issue, and when we start looking into these work items (possibly as we are doing P2P) see what are our options and what we need to do..

**Comment by stonezhong** [comment]
I know this is not javascript, but I made a tool to deal with similar issue in javascript:
if you have a promise library, you can deal with it much easier with /** @async **/ functions, you call these functions as if they are sync-functions, but underneath, the async-compiler-runtime will try to solve the asynchronous. I am adding some feature so it allow multiple promise can run in parallel so it does not lose performance.

https://github.com/stonezhong/async-compiler

**Comment by jpsfs** [comment]
Any news on this?
As larger projects surface more and more, I think we are desperately needing this!

**Comment by justinfagnani** [comment]
Also wondering about updates. I just hit this while trying to write a CompilerHost for a system that only has async file reading APIs.

**Comment by mhegazy** [comment]
This is not something in our immediate plans at the time being.

**Comment by felixfbecker** [comment]
This issue hasn't been active for a while, but I wanted to share our use case nevertheless.
I am working on the TypeScript support for Sourcegraph. You can explore code online with code intelligence, eg https://sourcegraph.com/github.com/Microsoft/TypeScript.

Initialization time (the time from the visit of the page and the first showing hover) is a high priority for us, and atm it definitely is not where we want it to be. For something simple as a hover over a local symbol in a file, we need to do a full checkout of all the TypeScript files in the repo, read them into memory just to be able to provide them synchronously - no matter what files are actually needed. This is exactly what @CyrusNajmabadi described for VS in https://github.com/Microsoft/TypeScript/issues/1857#issuecomment-99315295. 

> To add to what Cyrus said, I think it would generally be helpful to see some data regarding this before we came to any sort of conclusion - both on the compiler side as well as the LS side.

I can ensure you that fetching all files from an external git service instead of on-demand per-file is a huge difference. Installation of dependencies also counts into this, since again, we _might_ need these files (we all know how long a fresh npm/yarn install can take). I would be happy to provide concrete numbers if that helps.

>  what I am saying is this should be done as part of other work items. We need to do scalability and performance work items, but these are not scheduled.. So all I am saying is close this issue, and when we start looking into these work items (possibly as we are doing P2P) see what are our options and what we need to do..

all of our language support is backed by the language server protocol, and we currently [use one that wraps the TS service]([https://github.com/sourcegraph/javascript-typescript-langserver]). Having this would make implementing LSP (#11274) almost trivial.

**Comment by cancerberoSgx** [comment]
Any news on this ? Right now Language Service only allows non-interactive refactors. I think that's the cause why nobody has made interesting refactors like "move class to another file" or "move member to another class". In these cases you need to ask the user the target destination and you cannot do it if methods like getEditsForRefactor() are sync. I think this is important. Maybe I'm missing something? Thanks!

**Comment by belohlavek** [comment]
Also wondering about updates. I just hit this while trying to write a CompilerHost with a virtual/browser based FS.

**Comment by felixfbecker** [comment]
We have async/await now, which allows us to write async code that is almost the same syntax as sync code. What if we just changed `readFile` et al to be `async`, and put `await` before every call? There would be no difference in behavior, but the thread would not be blocked while doing IO and contents could be supplied through async ways like over the network.

I experimented with doing this manually for `readFile`: https://github.com/Microsoft/TypeScript/compare/master...felixfbecker:async-experiment
It was actually pretty easy because TypeScript even has a code fix for making a function async. I did not make anything parallel, just converted the functions to make sure the behavior is 1:1 the same.
It does highlight very nicely however which functions are actually potentially doing IO without one suspecting them to, and a lot of places where IO _could_ be parallelised for better speed.

What does the TS team think of this?

**Comment by cancerberoSgx** [comment]
@felixfbecker Nice! do you know if actual LanguageService methods like getCompletionsAtPosition, getEditsForRefactor supports this too ? If so it would be a cheap PR since as I understand it only imply to change the signatures not the implementation right ? Thanks

**Comment by CyrusNajmabadi** [comment]
Note: asynchrony here is *very* tricky.  This is because hte language service and compiler work over *mutable* data structures (for example, the syntax trees in source files).  So, you have to be extremely careful that your async calls don't mean you come back and the world is now in a different state than you expect.

The only place async/await could potentially work would be in the initial sync-up between the LS and the host.  And even then, we wouldn't want to allow htat to be interrupted with our sync-up calls.    Once hte LS starts processing though, it basically should not be preempted with asynchrony.

**Comment by cancerberoSgx** [comment]
@CyrusNajmabadi thanks, but a question, when you said "asynchronicity here", "here" is referring to every interfaces like CompilerHost, LanguageServiceHost (the ones that perform FS) or  to LanguageService only ? In other words, is the @felixfbecker suspiction that FS access interally support async functions out of the box althouh this is not formalized in the signatures ?

**Comment by CyrusNajmabadi** [comment]
> s the @felixfbecker suspicion that FS access internally support async functions out of the box although this is not formalized in the signatures ?

I think the question is a little too broad for me to be able to take a bite of that.  Basically, all i'm saying is:

Asynchrony presents concurrency/reentrancy concerns that you have to be super careful with.  Specifically, at any point where you 'await', you may have arbitrary code run before the 'await' returns'.

This is highly problematic for the LS without being super careful.  Specifically, consider how any LS operation works.  Something calls into the LS, then the LS attempts to synchronize with the host *mutating shared state* as it does so.  If, during synchronization, an asynchronous call happens, then someone else may come in and mutate that shared state.  

This is a *solvable* problem.  In general, by going with the work-queue model.  In other words, any LS operation that comes in is put into a queue and only one operation can come off that queue and operate (even asynchronously) at at time.  It just needs to be done carefully and conscientiously.  Often times, many projects can jsut move to async/await, virally extending out the async APIs as necessary.  This isn't the case here.  Because of shared mutable state, special care needs to be taken (somewhat equivalent to how one might need to lock in multi-threaded systems).

**Comment by ahnpnl** [comment]
I hope this will get prioritized

**Comment by randfur** [comment]
> This is highly problematic for the LS without being super careful. Specifically, consider how any LS operation works. Something calls into the LS, then the LS attempts to synchronize with the host mutating shared state as it does so. If, during synchronization, an asynchronous call happens, then someone else may come in and mutate that shared state.

If you await all the way back up the call stack then there shouldn't be any introduction of possible race conditions. There should be no "If, during synchronization, an asynchronous call happens" because everything will be awaiting the readFile() just like in the synchronous code.

**Comment by CyrusNajmabadi** [comment]
> because everything will be awaiting the readFile() just like in the synchronous code.

How do you ensure that?

**Comment by nukisman** [comment]
Hey folks! What is the state of this feature?

PS: Make the Compiler great again!

**Comment by craigphicks** [comment]
### 1. Does a canonical transformation of sync to async which guarantees same order of execution exist?

To clarify the meaning of "same order of execution", I mean the same order of execution of that code which modifies/reads external objects. 

@CyrusNajmabadi 

> Asynchrony presents concurrency/reentrancy concerns that you have to be super careful with. Specifically, at any point where you 'await', you may have arbitrary code run before the 'await' returns'.

I am not convinced that is true.  Let us suppose that, in the case of a single top level function that calls multiple lower single level functions, after conversion to async, and properly `await`ing each sub function and making the appropriate changes to the form of the exception handling, the order of code execution will be exactly the same. (I believe this to be a property of concurrent programming, one that does not hold for parallel programming.)  Then I would argue that it can be proven by induction that execution order would be the same for any program.  

For the remainder of this discussion, I will assume that such a "safe" transformation exists in Javascript, because it is a concurrent language.  However, if that it wrong, then the remainder of the discussion is on shaky ground.
 
### 2. If such canonical conversion exists, it must be automated to be useful.

For the sake of argument, let us suppose  that under some canonical transformation to async that guarantees same order of code execution were guaranteed.  Then there is still the issue of achieving that canonical transformation.  Doing that manually is time consuming and bug prone.  If it were done by humans,  it could take months for some selected sync version to be translated to async and debugged.  For one thing, all the generic array iterator functions, (`forEach`, `map`, etc.)  have to be converted to `for` loops.  (It so happens that a coding rule of the Typescript project is to use generic array iterator functions whenever possible, so there are a lot of those.  

Therefore it seem safe to assume that an asynchronous version will never be implemented unless automated transformation by software is available.  Progress to date supports that assumption. 

So obviously such transformation to a canonical async form should be done by automation, if possible.  Is it possible?  It seems to me that automated transformation to async should be possible.   

### Scope of transformation

@DanielRosenwasser 

> This means that at the very least, the following would change to have asynchronous counterparts.
>
>    createSourceFile (because of /// <reference ... /> comments)
>    createProgram
>    all of the language service API


Right, it is not necessary to change every function to async.  
Most of the selection can be done automatically:
- 1.  Manually select the bottom-most target functions, e.g., file IO functions.  (Is there anything else?  Maybe some inter-process communication, calls to Worker threads?) 
- 2.  Using an automatically generated call tree map, work up the tree the from the bottom-most marked functions, marking each parent along the way for selection.  

The collection of all selected function are targeted for transformation. 

###  Making use of the transformed version

A transformation to async which guarantees same order of execution is not going to be any faster than the original.
The advantage only happens when the transformed version is modified to run multiple IO operations in concurrence (!== "in parallel").  Here, some care has to be taken.  For example, if we inform another external agent "I have written file/buffer X, you can read it now", we have to make sure the write has completed first.  But since we were already doing that originally, it doesn't seem hard to maintain.

As the async version is gradually and manually optimized, the original version will be undergoing gradual version improvements.

- Let T(N) be typescript version N
- Let AsyncT(N)  be the automatically transformed verison of T(N)
- Let AsyncT(N)-optimized be the manually modified version of AsyncT(N) where some sub function calls are now being grouped to run concurrently. 

When a new synchronous version T(N+1) of Typescript is prepared, an automated transformation from T(N+1) to AsyncT(N+1) is generated.  For the most part, git merging can be used to merge AsyncT(N)-optimized and AsyncT(N+1) into 
AsyncT(N+1)-optimized.  There might be some small amount of manual attention required (but infinitely less than if automated transformation were not used).   

### Support of synchronous API

Furthermore, there are clients of the current API who depend upon the synchronous API.  So those versions have to be maintained.  You have to decide whether to maintain two separate versions of the API, one having some async calls, and the other having no async calls, or one version of the API contain two differently named versions of some calls, e.g., `createProgram` and `createProgramSync`.  I think having two separate API versions is a better solution, so that the asynchronous one is a pure canonical transformation of the synchronous one.   Eventually the pure-synchronous version could be pulled. 

### Conclusion

In conclusion, a clean path to leveraging asynchronous IO exists, provided the following conditions are satisfied:
1.  It is true that any sync ts program can be transformed to a "safe" async ts program, where "safe" is as described in section 1. (As stated earlier, I believe this an inherent property of concurrent languages.)
2. That transformation  can be done automatically.  (Of course it can!)
3. Software to do said automatic transformation exists.
   
(I'd love to work on that project if I could find a life support system to make that possible.)

**Comment by CyrusNajmabadi** [comment]
@craigphicks I don't really understand the argument that is being made.  All i'm saying is this:

The TS compiler/language-server is heavily mutable (by design).  The current design presumes that any mutation is safe because it happens, and then all processing that needs to view that mutation will run after it all teh way to completion.  In other words, computation that operates on that mutable state will not ever see that mutable state change while it executes.  

If we allow for asynchrony here in operations, then this assumption may get blown out the window.  A computation operation may yield, allowing another operation to start.  That operation may mutate the shared state, which the old operation will now see when it resumes execution.  That will violate invariants of the system and can easily lead to corruption and/or crashes/data-loss.  

--

In no way am I saying that TS cannot become async.  What i'm saying simply is that it is not trivial and would need to be done witha  very conscious eye to ensure that either shared state mutations do not happen.  Or, if they do, they cannot cause problems to in-flight operations.

**Comment by CyrusNajmabadi** [comment]
> all of the language service API

If any LS api becomes async, that would allow another LS api to operate concurrently with it.  An LS API request can and will mutate the shared state of the LS server.  As such, if an LS api yields, and another LS api then runs, it can mutate shared state that will break the first LS api call when it resumes.

**Comment by justinfagnani** [comment]
I don't think the line between sync and async is so stark in this case.

While it's true that while the compiler is waiting on I/O, something else might alter program state it turns out that this is already true with a synchronous compiler host - the host can already modify program state in the existing sync callbacks. That would maybe be unusual, but there's certainly no existing guarantees here in any place where the compiler calls out to user code, and those callouts are precisely the places that need to be made async.

**Comment by CyrusNajmabadi** [comment]
> the host can already modify program state in the existing sync callbacks.

No it can't.  Because it is sync (and JS is single threaded) the host can't make any changes as the event thread is blocked progressing whatever sync operation is in flight.

> That would maybe be unusual, but there's certainly no existing guarantees

Yes there are.  We don't modify shared mutable state once an LS computation operation starts.  That's a necessary current invariant on how the system is built.  We would either need to find some way to keep this invariant in an async world.  Or we'd need to update all our features to support the idea that the shared state they are examining may mutate across an async yield point.

**Comment by justinfagnani** [comment]
This isn't about single or multi threading - the sync callback itself, say `readFile` or `getScriptSnapshot ` could modify state. These are user-written callbacks and can do dangerous mutations synchronously. Any call to a user-provided callback is just as dangerous as an await.

**Comment by DanielRosenwasser** [comment]
I think Cyrus is talking more about the risks of having to be very intentional in preventing state modification. An arbitrary host can create a language service, bork the state, and if an API consumer filed an issue saying "it hurts when I do this", much of the time we'd probably say "don't do that".

On the other hand, if you allow the API to re-enter asynchronously, that becomes a much more deliberate API commitment; I think that's probably half the use-case of making the API asynchronous. Even if we tell people "don't do that", it's way too easy for API consumers to accidentally re-enter the API.

**Comment by CyrusNajmabadi** [comment]
> This isn't about single or multi threading - the sync callback itself, say readFile or getScriptSnapshot could modify state.

How?  These calls just return values that the LS uses to update it's own state.  That state is then able to be used for the remainder of hte computation.

> any call to a user-provided callback is just as dangerous as an await.

It really isn't.   Calls to user-provided callbacks don't generally make reentrancy happen.  Calls to awaits do.

**Comment by craigphicks-public** [comment]
> @craigphicks I don't really understand the argument that is being made. All i'm saying is this:
> 
> The TS compiler/language-server is heavily mutable (by design). The current design presumes that any mutation is safe because it happens, and then all processing that needs to view that mutation will run after it all teh way to completion. In other words, computation that operates on that mutable state will not ever see that mutable state change while it executes.

It get what you are saying - that there are both multi-part read and write operations which must take place atomically - at least atomic writings cannot overlap atomic readings or other atomic writings, although atomic readings can overlap each other without issue. 

Moreover, it might be the case that those multi-part atomic write operations in LS include calls to `createProgram` and `emit`, so that asynchronous calls to IO under `createProgram` and `emit` could result in violating that atomicity.

How about using context (LS vs non-LS) to conditionally call async-IO under `createProgram` and `emit`?  A hack to enable some optimization until LS atomicity was enabled.

On the subject of atomic writes and reads, that's an issue that atomic-RW supporting databases handle.  Perhaps we could avoid reinventing the wheel and use such a database as an intermediary for the language service?

**Comment by craigphicks** [comment]
> @craigphicks I don't really understand the argument that is being made. All i'm saying is this:
> 
> The TS compiler/language-server is heavily mutable (by design). The current design presumes that any mutation is safe because it happens, and then all processing that needs to view that mutation will run after it all teh way to completion. In other words, computation that operates on that mutable state will not ever see that mutable state change while it executes.

It get what you are saying - that there are both multi-part read and write operations which must take place atomically - at least atomic writings cannot overlap atomic readings or other atomic writings, although atomic readings can overlap each other without issue. 

Moreover, it might be the case that those multi-part atomic write operations in LS include calls to `createProgram` and `emit`, so that asynchronous calls to IO under `createProgram` and `emit` could result in violating that atomicity.

How about using context (LS vs non-LS) to conditionally call async-IO under `createProgram` and `emit`?  A hack to enable some optimization until LS atomicity was enabled.

On the subject of atomic writes and reads, that's an issue that atomic-RW supporting databases handle.  Perhaps we could avoid reinventing the wheel and use such a database as an intermediary for the language service?

**Comment by CyrusNajmabadi** [comment]
> On the subject of atomic writes and reads, that's an issue that atomic-RW supporting databases handle. Perhaps we could avoid reinventing the wheel and use such a database as an intermediary for the language service?

i see no way to do that without enormous complexity.  TS is not Roslyn.  It cannot trivially fork the state of the LS (esp. given the perf constraints of JS engines).  So it has to take a direct mutation approach just to maintain reasonable perf (and even with that, perf is a continual issue).  TS's design is heavily predicated on shared mutation.  So either we have to: 

1. ensure that shared mutation cannot happen with asynchrony (not sure how). or
2. ensure that shared mutation doesn't impact resumed computation (hard, but possible).  or
3. redesign not based on a shared mutation model.  extremely cost.  underpins the entire design of hte TS compiler/LS.

**Comment by craigphicks** [comment]
> > On the subject of atomic writes and reads, that's an issue that atomic-RW supporting databases handle. Perhaps we could avoid reinventing the wheel and use such a database as an intermediary for the language service?
> 
> i see no way to do that without enormous complexity. TS is not Roslyn. It cannot trivially fork the state of the LS (esp. given the perf constraints of JS engines). So it has to take a direct mutation approach just to maintain reasonable perf (and even with that, perf is a continual issue). TS's design is heavily predicated on shared mutation. So either we have to:
> 
>     1. ensure that shared mutation cannot happen with asynchrony (not sure how). or
> 
>     2. ensure that shared mutation doesn't impact resumed computation (hard, but possible).  or
> 
>     3. redesign not based on a shared mutation model.  extremely costly.  underpins the entire design of hte TS compiler/LS top to bottom.

My thinking may to too simple, or maybe the overhead is too high.  Inside LS, an javascript-`async` should-be-atomic multi-part write op calls a worker to make a synchronous write to the database and awaits the workers return.  The use of the database ensures that javascript-`async` should-be-atomic multi-part reads do not see a half written version.

**Comment by CyrusNajmabadi** [comment]
> to make a worker-thread-synchronous write to the database 

What database are you referring to?  We have no databases here in our impl.  As stated above, we have a shared-memory, mutable-state, representation of everything.  

> The use of the database ensures that javascript-async should-be-atomic multi-part reads do not see a half written version.

So this is either 1 or 3 (closer to 3 from my reading).  It would be an entire redesign of how everything works from *top to bottom*.  We'd need to introduce a DB indirection layer into the deepest parts of our system, which have been tuned heavily for perf.  There is no indication whatsoever that this could possibly work or meet any sort of perf constraints we have.

**Comment by CyrusNajmabadi** [comment]
>  The only requirement it satisfies is ensuring transactional consistency,

This is a huge requirement.  It's effectively stating: move away from a shared-state/mutable system as our backing store.  Yes, that would potentially address things.  But it would need an effective full rewrite of major parts of teh stack.  And there's no data at all that indicates that this could be implemented in an efficient enough fashion.  

If we were to contemplate a (near) full rewrite, i'd rather just go the roslyn approach and have an shared-state/immutable (rather than mutable) model.

**Comment by craigphicks** [comment]
> > The only requirement it satisfies is ensuring transactional consistency,
> 
> This is a huge requirement. :)

Unwind my database detour that would not be practical nor easier.

Here is a plan:

### protocol for ensuring atomic transaction consistency

At the heart of the matter is the protocol for ensuring atomic transaction consistency.  

The per file interface could look like 
```
interface Gatekeeper1 {
    async waitForReadPermission():void
    async readDone():void
    async waitForWritePermission():void
    async writeDone():void
}
```

or something that shifted the burden of waiting before execution to the gatekeeper such as 

```
type Awaited<T> = T extends PromiseLike<infer U> ? Awaited<U> : T;

interface Gatekeeper2 {
  async waitUntilRead<F extends (...args: any[]) => ReturnType<F>>(anyf: F)
    : (...args: Parameters<F>)=>Promise<Awaited<ReturnType<F>>>
  async waitUntilWritten<F extends (...args: any[]) => ReturnType<F>>(anyf: F)
    : (...args: Parameters<F>)=>Promise<Awaited<ReturnType<F>>>
}
```

which would return a function that managed holding onto the arguments before calling the function, and also handle the later `...Done()` call, while still ensuring type checking on the arguments.  That would make it easier on the caller.


The instantiation of the interface would use a couple of queues to keep track of requests, one queue each for reads and writes, with writes taking precedence.  Also, it would keep track of the number of currently executing reads/writes.  For the next request to start,
 - a write request can start when the number of currently executing reads and writes are both zero.  
 - a read request can start when the number of currently executing writes are zero (and the write request queue is empty).
 
The gatekeeper logic is activated (released from an await state by a resolve) on every event -  i.e., a new request or a currently executing read or write completing. It's not even necessary to use nodejs message - it can be more transparently with promises managed by gatekeeper.
 

 ### Careful development and testing
 
1. Select a small handful of atomic multi-read / atomic multi-write candidates for testing.

2. Rewrite the current sync code so that each each atomic read / atomic write in the candidate group is callable as a function.  Call the group of function T.

3. Use call tree analysis to identify all functions above T in the call tree and call that entire group ET.

4. Use the automated sync->async transform software to transform each function in transform ET to async form. Now async is available for coding in our targets T. 
 
5. Set up gatekeeper protocol for each call to a function in T.

6. Within each function in T, change the file-IO functions from sync to async.

7. Unit testing, integrated testing.  

8. Back to 1. to add a few more and repeat (more or less).

Note:  The automatic transform of sync to async creates a huge amount of code changes across a wide range of code.  For the most part we don't have to worry about that, e.g., if the sync version master undergoes drastic change, then a new auto async version can be generated.  We only have worry about merging the code of functions in T and the immediate callers of functions in T into that new auto-generated version. 


### Conclusion

It is possible to break the problem down into small steps allowing normal development.  

That is true despite the fact that 
1.  lexically speaking, a very large number of functions that do not seem directly related to our goal have to be transformed from sync to async form.  
     - That assessment is conditioned on the existence of software to automate sync to async format change for selected parts of the code.  Because that will allow humans to focus only on small areas of codes which require human attention. 
 2. Async is more complex than sync.   
    - Absolutely true.  However, there are rules and mechanisms to get the job done, provided that incremental development and testing is followed, so that manually estimated should-be-atomic read/write groupings can be coded and immediately tested.  Concurrency takes the headache out of paralllel processing and leaves only the fun part.

**Comment by CyrusNajmabadi** [comment]
That feels like an effective rewrite of the TypeScript compiler/LS , along with teh development of *several* new systems and tools needed to support the sync/async hybrid model.  

It's also unclear how any of this is enforced.  It seems to be a cooperative reader/writer lock approach.  But a single errant yield means mutations may happen when problematic.  In a codebase the size and complexity of TS, i don't see how that is prevented.

**Comment by craigphicks** [comment]
> It seems to be a cooperative reader/writer lock approach

Enforced inside the library.  The library clients are not required to do anything special just because they are making re-entrant calls.  

One thing we didn't talk about is any variables that might be in the a shared top level scope between re-entrant LS calls, but that really should belong to only one calls scope. 

About how many separate must-be-atomic multi-read and/or multi-write operations do you estimate there are?  

> In no way am I saying that TS cannot become async. What i'm saying simply is that it is not trivial and would need to be done witha very conscious eye to ensure that either shared state mutations do not happen. Or, if they do, they cannot cause problems to in-flight operations.

Exactly right,  I agree.

**Comment by craigphicks** [comment]
@CyrusNajmabadi 
> That feels like an effective rewrite of the TypeScript compiler/LS , along with teh development of several new systems and tools needed to support the sync/async hybrid model.

I feel strongly the the current sync libraries and the async libraries should be kept separate.  All version improvements other than the async conversion changes should be applied to the sync version, and the async conversion branch updated from the sync version. 

A client of LS would use one or the other.    I don't see any way to mix them for LS.  I don't see any reason to mix them for compiler API.  *If async performance gains were sufficient*, adoption would naturally increase, especially new projects.  At some point, non-async version up support could be dropped.

I would guess that async is not the only bottleneck. It sometimes seems the compiler is loading a lot a files housing symbols which are never ever actually referenced during `createProgram`, and at least in debug mode this noticeably slows things down.   If the file loading occurred just-in-time when the symbol was needed it might speed things up in some cases (at least debug mode).  That comment is off topic here - I don't claim that excessive file loading is really a critical bottleneck - just mentioning it as a possible way in which async might not be the most critical bottleneck, although it might be an important precursor to solving other bottlenecks.

**Comment by partic2** [comment]
Automated transformation (from sync to async) sounds like a good idea, It's there any tools/project can do or trying to do this?

**Comment by RyanCavanaugh** [comment]
In TS7, API will be out of proc thus up to the caller to decide synchronicity (in-proc, if you must, will have Go non-coloring semantics, thus this is moot).
